src/test/java/com/eighthlight/fabrial/test/client/EchoRequest.java

```diff
@@ -0,0 +1,41 @@
+package com.eighthlight.fabrial.test.client;
+
+import java.io.*;
+import java.util.logging.Logger;
+
+public class EchoRequest {
+  OutputStream outputStream;
+  InputStream inputStream;
+
+  Logger logger;
+
+  public EchoRequest(OutputStream outputStream,
+                     InputStream inputStream) {
+    this.outputStream= outputStream;
+    this.inputStream = inputStream;
+    this.logger = Logger.getLogger(this.toString());
+  }
+
+  public <T extends Serializable> T send(T object) throws IOException, ClassNotFoundException {
+    write(object);
+    return read();
+  }
+
+  private <T extends Serializable> void write(T object) throws IOException {
+    try (ByteArrayOutputStream bytesOut = new ByteArrayOutputStream();
+        ObjectOutput objectOut = new ObjectOutputStream(bytesOut)) {
```

Comment:
I see you're using the native Java serialization APIs (e.g. `readObject`) throughout this class - I'd suggest reconsidering, since you really just want to echo raw bytes back, not necessarily serialized Java objects. You should be able to just read/write bytes from the stream as long as they keep coming (or up to a newline, or however you want to delimit it). Nothing in the server project will need Java serialization.

Reply:
> since you really just want to echo raw bytes back, not necessarily serialized Java objects

True, and I might be guilty of some bike-shedding and not adhering to YAGNI. However, I was also approaching this exercise as an opportunity to learn idiomatic, modern Java patterns. Even still, this API is arguably over-engineered for a test class that isn't in the main vein of the project trajectory (i.e. HTTP). I'll try to keep this in mind going forward.

Reply:
👍 Yep, understood. For what it's worth, for modern Java serialization, I'd tend to steer clear of the native `Serializable` APIs. For various use cases I'd look at one of JSON/XML, Avro, Protocol Buffers, or even Kryo.

Reply:
Good to know